litt/disktable/flush_coordinator.go

```diff
@@ -0,0 +1,151 @@
+	alive atomic.Bool
+}
+
+// A request to flush the underlying database. When the flush is eventually performed, a response is sent on
+// the request's channel. The response is nil if the flush was successful, or an error if it failed.
+type flushCoordinatorRequest chan error
+
+// Creates a new flush coordinator.
+//
+// - internalFlush: the function that actually performs the flush on the underlying database
+// - flushPeriod: the minimum time period between flushes, if zero then no batching is performed
+func newFlushCoordinator(
+	errorMonitor *util.ErrorMonitor,
+	internalFlush func() error,
+	flushPeriod time.Duration,
+) *flushCoordinator {
+
+	fc := &flushCoordinator{
+		errorMonitor:  errorMonitor,
+		internalFlush: internalFlush,
+		requestChan:   make(chan any, requestChanBufferSize),
+	}
+	fc.alive.Store(true)
+
+	if flushPeriod > 0 {
+		fc.rateLimiter = rate.NewLimiter(rate.Every(flushPeriod), 1)
+		go fc.controlLoop()
+	}
+
+	return fc
+}
+
+// Flushes the underlying database. May wait to call flush based on the configured flush period.
+func (c *flushCoordinator) Flush() error {
+	if c.rateLimiter == nil {
+		// we can short circuit and just call the internal flush directly, flush frequency is infinitely high
+		return c.internalFlush()
+	}
+
+	request := make(flushCoordinatorRequest, 1)
+
+	// send the request
+	err := util.Send(c.errorMonitor, c.requestChan, request)
+	if err != nil {
+		return fmt.Errorf("error sending flush coordinator request: %w", err)
+	}
+
+	// await the response
+	response, err := util.Await(c.errorMonitor, request)
+	if err != nil {
+		return fmt.Errorf("error awaiting flush coordinator response: %w", err)
+	}
+
+	if response != nil {
+		return fmt.Errorf("flush failed: %w", response)
+	}
+	return nil
+
+}
+
+// Closes the flush coordinator, cleaning up any resources.
+func (c *flushCoordinator) Close() {
+	c.alive.Store(false)
+}
+
+// The control loop that manages flush timing.
+func (c *flushCoordinator) controlLoop() {
+	defer close(c.requestChan)
+
+	// requests that are waiting for a flush to be performed
+	waitingRequests := linkedlistqueue.New()
+
+	// timer used to wait until the next flush can be performed
+	timer := time.NewTimer(0)
+	var timerActive bool
+
+	for c.alive.Load() {
+
+		if timerActive {
+			// There are pending flushes we want to handle, but we need to wait until the timer expires.
+			select {
+			case <-c.errorMonitor.ImmediateShutdownRequired():
+				return
+			case request := <-c.requestChan:
+				waitingRequests.Enqueue(request)
+			case <-timer.C:
+
+				// we can now perform a flush
+				err := c.internalFlush()
+				// send a response to each waiting caller
+				for request, ok := waitingRequests.Dequeue(); ok; request, ok = waitingRequests.Dequeue() {
+					request.(flushCoordinatorRequest) <- err
+				}
+
+				timerActive = false
+			}
+		} else {
+			// There are pending flushes, we can handle requests immediately if the rate limiter allows it.
```

Comment:
Not sure I understand what this comment is getting at. Do you mean there are NO pending flushes? Or something else?

Reply:
I agree this comment is confusing. Updated to the following:

```
			// We don't have any pending flush requests, so we aren't waiting on the timer. If we get a new request,
			// check to see if the rate limiter will allow it to be flushed immediately, and do so if possible.
```